agent/consul/fsm/commands_oss.go

```diff
@@ -50,6 +50,9 @@ func (c *FSM) applyRegister(buf []byte, index uint64) interface{} {
 		c.logger.Printf("[WARN] consul.fsm: EnsureRegistration failed: %v", err)
 		return err
 	}
+
+	c.publisher.PublishEvent(&req, index)
```

Comment:
Interesting. I was thinking that we'd integrate publishing directly into the `state` package as part of the memDB transaction eventually.

this is simpler for now but I wonder how far it will get us in cases where we need more information that is present in the raft log entry.

In fact I think it's already an issue here - for deregister below the `req` doesn't have enough information to properly encode a topic etc. which is the whole reason we can't just use the raw raft logs in the first place.

So I think this all needs to move up to be in the `state` package rather than the FSM layer since the FSM doesn't "know" enough about the current state to be able to rebuild the info we need.